compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIECreateAIEWorkgroup.cpp

```diff
@@ -70,24 +70,58 @@ LogicalResult workgroupBuildForDmaCpyNdOp(
     Block::iterator controlCodeBegin, Block::iterator controlCodeEnd) {
   LLVM_DEBUG(llvm::dbgs() << "workgroupBuild [amdaie.dma_cpy_nd] Start\n");
   Attribute sourceMemSpace = dmaOp.getSourceObjectFifo().getMemorySpace();
+  Attribute targetMemSpace = dmaOp.getTargetObjectFifo().getMemorySpace();
   Location loc = rewriter.getUnknownLoc();
   SmallVector<OpFoldResult> empty;
+
+  SmallVector<OpFoldResult> circularDmaTargetOffsets = empty;
+  SmallVector<OpFoldResult> circularDmaTargetSizes = empty;
+  SmallVector<OpFoldResult> circularDmaTargetStrides = empty;
+  SmallVector<OpFoldResult> circularDmaSourceOffsets = empty;
+  SmallVector<OpFoldResult> circularDmaSourceSizes = empty;
+  SmallVector<OpFoldResult> circularDmaSourceStrides = empty;
+
+  SmallVector<OpFoldResult> ipuDmaTargetOffsets = dmaOp.getTargetMixedOffsets();
+  SmallVector<OpFoldResult> ipuDmaTargetSizes = dmaOp.getTargetMixedSizes();
+  SmallVector<OpFoldResult> ipuDmaTargetStrides = dmaOp.getTargetMixedStrides();
+  SmallVector<OpFoldResult> ipuDmaSourceOffsets = dmaOp.getSourceMixedOffsets();
+  SmallVector<OpFoldResult> ipuDmaSourceSizes = dmaOp.getSourceMixedSizes();
+  SmallVector<OpFoldResult> ipuDmaSourceStrides = dmaOp.getSourceMixedStrides();
+  if (!sourceMemSpace) {
+    // Check if the source of DmaCpyNd op is from L3 - then addressing will be
+    // controlled by the uController.
```

Comment:
Suggestion for further clarification:

```suggestion
    // Check if the source of DmaCpyNd op is from L3 - then source addressing
    // will be controlled by the uController and target addressing will stay in the
    // circular DMA to be part of the AIE configuration.
```